Inventory: Added new parser to allow nested expressions

mliu01 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
_I haven't contributed code like this before, so if there's anything I have to consider, please let me know! :)_

Added a new parser for patterns / selecting hosts.
This parser allows nested patterns such as `host1:((host2:&host3):!host4)` to freely select hosts with just one command when using `--limit`, within playbooks under `hosts: ` and it should work with lookup plugins too.

It uses the [pyparsing ](https://github.com/pyparsing/pyparsing)library to specify expression grammar and supports nested patterns containing any combination with:
- all hosts with `all`
- indexing, e.g. `hostgroup[1]`
- plain host 
- hostgroup
- ipv4 and ipv6 addresses (also correctly parses ipv6 addresses in combination with the double colon `:` OR operator)

All other functionalities Ansible patterns has [as listed here](https://docs.ansible.com/ansible/2.8/user_guide/intro_patterns.html) are still supported.
The syntax is extended by a few other characters, but you can still use the old syntax without any issues. 

**AND** operators include: `&& & :&`  `#Intersection of two groups; host has to be in both groups`
**OR** operators include: `, : `  `#Includes specified host`
**NAND** operator: `:!`  `#Intersection of left host and not right host; added to keep ansible pattern syntax the same`    
**NOT** operators include: `! !!` `#Excludes specified host; for NAND or NOR usage: needs to be in combination with either AND or OR operators)`   

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
When selecting hosts that are saved in multiple groups, you would have to call a playbook multiple times with `--limit` to get the ones you specifically are asking for, since nested patterns are ignored and not allowed .

The `InventoryManager` is extended by another class called `PyparsingPatternParser`, which is responsible for the parsing and includes a few other ParseAction functions (features from pyparsing) that select the hosts from inventory.
Evaluates in this order: **NAND > NOT > AND > OR** 
_(You can use parentheses to change the order to your liking.)_

Another class called `AnsiblePatternParser` contains all the functions used to parse pattern the usual way. If the pyparsing library cannot be imported or if the Python version is below 3.6, this parser will be used instead.

<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
**hosts file**
````
[hostgroup1]
host1
host2
host3

[hostgroup2]
host2
host3
host4

[hostgroup3]
host4
host5

[hostgroup4]
host6
````


**before**
```
$ ansible-playbook -i hosts helloworld.yml --limit 'hostgroup4:(hostgroup2:&hostgroup3):!hostgroup1'
[WARNING]: Could not match supplied host pattern, ignoring: (hostgroup2
[WARNING]: Could not match supplied host pattern, ignoring: hostgroup3)
ERROR! Specified hosts and/or --limit does not match any hosts
```


**after**
```
$ ansible-playbook -i hosts helloworld.yml --limit 'hostgroup4:(hostgroup2:&hostgroup3):!hostgroup1'
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel"
if you are modifying the Ansible engine, or trying out features under development. This is a rapidly
changing source of code and can become unstable at any point.

PLAY [all] ***********************************************************************************************

TASK [shell] *********************************************************************************************
fatal: [host4]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname host4: Name or service not known", "unreachable": true}
fatal: [host6]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: ssh: Could not resolve hostname host6: Name or service not known", "unreachable": true}

PLAY RECAP ***********************************************************************************************
host4                      : ok=0    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0
host6                      : ok=0    changed=0    unreachable=1    failed=0    skipped=0    rescued=0    ignored=0
```

Commit:
Inventory: Added new parser to allow nested expressions

added warnings for ignoring patterns that do not yield matches

fixed implicit host und subset parsing

bcoca commented on lib/ansible/inventory/manager.py:
```diff
@@ -28,6 +28,7 @@
 
 from operator import attrgetter
 from random import shuffle
+from pyparsing import pyparsing_common, Word, Literal, Combine, Optional, ZeroOrMore, alphas, alphanums, nums, printables, opAssoc, infixNotation, oneOf
```
The problem with adding new dependencies is that it limits the systems on which ansible core is installable, if we could do same w/o adding this library it is a lot more probable that we can merge this (also clog and tests would be nice).

We do have a vendoring facility but that puts an onus on maintenance on the core team side so it has to be a very compelling reason to do so.

ciis0 replied:
(I am mliu01's technical supervisor for this project.)

What about making it an optional dependency, making it available only on supported systems?
My assumption here would be that systems where you can't install pyparsing might not need as sophisticated limits as machines where you can. Not sure how bad it is to have to implementations of `--limit`.

Or otherwise what about introducing an plugin facility for `--limit`? then the current/previous limit logic would be the default plugin and users can implement their own limit plugins.

bcoca replied:
Making it optional might work.

It does not affect just `--limit`, host patterns can be used in `hosts:` keyword and in a lookup.

I was trying to figure out how to put this into a custom inventory plugin but they are not designed to override the parsing/selection of hosts aka dealing with 'host patterns'.

ciis0 replied:
We'll try making it optional then. :)

bcoca commented on lib/ansible/inventory/manager.py:
```diff
@@ -46,9 +47,14 @@
+    b'|'.join(
+        IGNORED_ALWAYS +
+        IGNORED_PATTERNS +
+        IGNORED_EXTS))
```
there are also a bunch of style changes in this PR, i would remove those to limit the scope of what needs to be reviewed.

mliu01 replied:
oh man i must have formatted it again, just reverted the style changes with the newest commit!

Commit:
added HAS_PYPARSING variable for optional dependency

seperated parser into AnsiblePatternParser and PyparsingPatternParser

Commit:
added changelog

Commit:
added correct parsing for hostnames starting with a number

fixed sets losing order after intersect/union/difference

Commit:
removing pyparsing functions from pickle; support not there regarding library

Commit:
added support for unicode characters and added comments

Commit:
changed pyparsingpatternparsers _get_subscript to use split and apply_subscript

Commit:
added syntaxerror expection handling in import

changed warning to debug

added newline in changelog

added pyparsing import only for python version 2.8 and above

added different limit handling when inventory empty

Commit:
added inventory_pyparsing integration test

update changelog

mliu01 commented:
The PR should be ready for review now :)
I added a changelog (75645-new-parser-to-allow-nested-expressions.yml) and an integration test called inventory_pyparsing. 
I'm not sure how to limit the test just for environments where the PyparsingPatternParser is being used (ie python >= 2.8 and successful pyparsing library import). 
The condition part in the runme.sh file might be a little unreliable, so please let me know how I should modify it or the test as a whole!

s-hertel commented:
Since the presence of the library would change the execution of the playbook, this should be behind a toggle (and be opt-in). This might be worth bringing up at an IRC meeting since we're not sure we should add this complexity.

For public community meetings - Matrix: #meeting:ansible.im | IRC: #ansible-meeting

https://github.com/ansible/community/issues?q=is%3Aissue+is%3Aopen+label%3Acore+label%3Ameeting_agenda

jborean93 commented:
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.
Our reasons for this are:

   * This is not a feature we are interested in at this point in time
   * There has been no response to bringing it up on the meeting agenda
   * We don't want to add extra optional dependencies to Ansible

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!
